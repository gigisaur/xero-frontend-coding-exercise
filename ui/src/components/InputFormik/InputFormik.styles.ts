import styled from "styled-components";
import { Color, FontFamily } from "../../theme";

export const Label = styled.label`
    user-select: none;
    color: ${Color.DefaultText};
    font-size: 0.9rem;
`;

export const ErrorP = styled.p`
    color: ${Color.ErrorRed};
    padding: 0;
    margin: 0;
`;

interface IInputProps {
    hasErrors: boolean;
}

export const Input = styled.input<IInputProps>`
    width: 100%;
    border: 1px solid ${(props) => props.hasErrors ? Color.ErrorRed : Color.Border};
    background-color: white;
    border-radius: 0.25rem;
    line-height: 1.3em;
    font-size: 0.9rem;
    text-indent: 0.8rem;
    padding: 0;
    box-sizing: border-box;
    font-family: ${FontFamily};
    color: ${Color.DefaultText};
    height: 2.5rem;
`;

export const ErrorContainerDiv = styled.div`
    line-height: 0.9rem;
    padding-bottom: 1rem;
    padding-top: 0.2rem;
    font-size: 0.8rem;
`;

export const FieldContentContainerDiv = styled.div`
    display: grid;
    grid-template-rows: 1.25rem minmax(2.5rem, auto) minmax(2.1rem, auto);
    box-sizing: border-box;
`;

export const Span = styled.span`
    margin-left: 0.25rem;
    color: ${Color.ErrorRed};
`;
