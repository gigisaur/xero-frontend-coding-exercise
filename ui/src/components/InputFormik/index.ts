import { FormikProps } from "formik";

type InputType = "text" | "number";

export interface IInputFormikProps {
    formikProps: FormikProps<any>;
    label: string;
    name: string;
    required?: boolean;
    type: InputType;
}

export * from "./InputFormik";
