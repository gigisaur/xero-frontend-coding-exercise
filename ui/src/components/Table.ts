import styled from "styled-components";
import { Color } from "../theme";

export const Table = styled.table`
    border-collapse: collapse;
    width: 100%;
    border: 1px solid ${Color.TableBorder};
`;

export const Td = styled.td`
    padding: 0.5rem;
    color: ${Color.DefaultText};
    font-size: 0.9rem;
    border-bottom: 1px solid ${Color.TableBorder};
`;

export const Th = styled.th`
    padding: 0.5rem;
    border-bottom: 1px solid ${Color.Border};
    color: ${Color.DefaultText};
    opacity: 0.75;
    font-size: 0.8rem;
    text-transform: uppercase;
    text-align: left;
`;

export const Tr = styled.tr`
    &:nth-child(odd) {
        background-color: ${Color.TableBorder};
    }
`;
