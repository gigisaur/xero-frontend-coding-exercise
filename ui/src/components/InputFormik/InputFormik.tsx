import { Field, FieldProps } from "formik"
import * as _ from "lodash";
import React from "react"
import { IInputFormikProps } from ".";
import { strings } from "../../localization";
import { ErrorContainerDiv, ErrorP, FieldContentContainerDiv, Input, Label, Span } from "./InputFormik.styles";

export const InputFormik = ({ name, formikProps, label, required, type }: IInputFormikProps) => {
    return (
        <Field name={name}>
            {(props: FieldProps) => (
                <FieldContentContainerDiv>
                    <Label htmlFor={name}>
                        {label}
                        {required &&
                            <Span title={strings.Label_Required}>*</Span>
                        }
                    </Label>
                    <Input
                        {...props.field}
                        aria-describedby={`${name}_describedby`}
                        aria-invalid={formikProps.errors[name] !== undefined && formikProps.touched[name] === true}
                        aria-required={required}
                        hasErrors={formikProps.errors[name] !== undefined && formikProps.touched[name] === true}
                        id={name}
                        type={type}
                    />
                    <ErrorContainerDiv>
                        {(formikProps && formikProps.errors && _.get(formikProps.errors, name))
                            && (formikProps && formikProps.touched && _.get(formikProps.touched, name))
                            && <ErrorP id={`${name}_describedby`} aria-live="assertive" role="alert" >{_.get(formikProps.errors, name)}</ErrorP>}
                    </ErrorContainerDiv>
                </FieldContentContainerDiv>
            )}
        </Field>
    );
}
