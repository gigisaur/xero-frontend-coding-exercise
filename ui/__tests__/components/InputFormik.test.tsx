import { act, fireEvent, render, wait } from "@testing-library/react";
import { Formik, FormikProps } from "formik";
import { axe } from "jest-axe";
import "jest-styled-components";
import * as React from "react";
import { InputFormik } from "../../src/components/InputFormik";

const propsMock: any = {
    label: "Foo label",
    name: "name",
    type: "text",
}

test(`When rendered then the HTML matches this snapshot`, () => {
    const { unmount, container } = render(
        <Formik
            initialValues={{ name: "" }}
            // tslint:disable-next-line: no-empty
            onSubmit={() => {}}
        >
            {(formProps: FormikProps<any>) => (
                <form onSubmit={formProps.handleSubmit}>
                    <InputFormik {...propsMock} formikProps={formProps} />
                </form>
            )}
        </Formik>);

    expect(container).toMatchSnapshot();
    unmount();
});

test(`Given type number when rendered then input has type correctly set`, () => {
    const { unmount, getByLabelText } = render(
        <Formik
            initialValues={{ name: "" }}
            // tslint:disable-next-line: no-empty
            onSubmit={() => {}}
        >
            {(formProps: FormikProps<any>) => (
                <form onSubmit={formProps.handleSubmit}>
                    <InputFormik {...propsMock} formikProps={formProps} type="number" />
                </form>
            )}
        </Formik>);

    expect(getByLabelText("Foo label").getAttribute("type")).toEqual("number");
    unmount();
});

test(`Given required is true when rendered then required asterisk is shown`, () => {
    const { unmount, getByText } = render(
        <Formik
            initialValues={{ name: "" }}
            // tslint:disable-next-line: no-empty
            onSubmit={() => {}}
        >
            {(formProps: FormikProps<any>) => (
                <form onSubmit={formProps.handleSubmit}>
                    <InputFormik {...propsMock} formikProps={formProps} required />
                </form>
            )}
        </Formik>);

    expect(getByText("*"));
    unmount();
});

describe(`Given there are errors for the input`, () => {
    test(`Then the error message is shown`, async () => {
        const { unmount, getByLabelText, getByText, queryByText } = render(
            <Formik
                initialValues={{ name: "" }}
                // tslint:disable-next-line: no-empty
                onSubmit={() => {}}
                validateOnBlur
                validate={() => ({ name: "Test error message" })}
            >
                {(formProps: FormikProps<any>) => (
                    <form onSubmit={formProps.handleSubmit}>
                        <InputFormik {...propsMock} formikProps={formProps} required />
                    </form>
                )}
            </Formik>);

        expect(queryByText("Test error message")).toBeNull();

        act(() => {
            fireEvent.blur(getByLabelText(/Foo label/g));
        });

        await wait();

        expect(getByText("Test error message"));
        unmount();
    });

    test(`Then there are no accessibility violations`, async () => {
        const { unmount, getByLabelText, container } = render(
            <Formik
                initialValues={{ name: "" }}
                // tslint:disable-next-line: no-empty
                onSubmit={() => {}}
                validateOnBlur
                validate={() => ({ name: "Test error message" })}
            >
                {(formProps: FormikProps<any>) => (
                    <form onSubmit={formProps.handleSubmit}>
                        <InputFormik {...propsMock} formikProps={formProps} required />
                    </form>
                )}
            </Formik>);

        act(() => {
            fireEvent.blur(getByLabelText(/Foo label/g));
        });

        const results = await axe(container)

        expect(results).toHaveNoViolations();
        unmount();
    });
})

test(`When rendered then there are no accessibility violations`, async () => {
    const { container, unmount } = render(
        <Formik
            initialValues={{ name: "" }}
            // tslint:disable-next-line: no-empty
            onSubmit={() => {}}
        >
            {(formProps: FormikProps<any>) => (
                <form onSubmit={formProps.handleSubmit}>
                    <InputFormik {...propsMock} formikProps={formProps} />
                </form>
            )}
        </Formik>);
    const results = await axe(container)

    expect(results).toHaveNoViolations();
    unmount();
});
