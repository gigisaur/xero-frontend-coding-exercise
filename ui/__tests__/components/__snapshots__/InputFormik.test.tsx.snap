// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When rendered then the HTML matches this snapshot 1`] = `
.c1 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #191919;
  font-size: 0.9rem;
}

.c2 {
  width: 100%;
  border: 1px solid #DAE5EC;
  background-color: white;
  border-radius: 0.25rem;
  line-height: 1.3em;
  font-size: 0.9rem;
  text-indent: 0.8rem;
  padding: 0;
  box-sizing: border-box;
  font-family: Roboto,Arial,Helvetica,sans-serif;
  color: #191919;
  height: 2.5rem;
}

.c3 {
  line-height: 0.9rem;
  padding-bottom: 1rem;
  padding-top: 0.2rem;
  font-size: 0.8rem;
}

.c0 {
  display: grid;
  grid-template-rows: 1.25rem minmax(2.5rem,auto) minmax(2.1rem,auto);
  box-sizing: border-box;
}

<div>
  <form>
    <div
      class="c0"
    >
      <label
        class="c1"
        for="name"
      >
        Foo label
      </label>
      <input
        aria-describedby="name_describedby"
        aria-invalid="false"
        class="c2"
        id="name"
        name="name"
        type="text"
        value=""
      />
      <div
        class="c3"
      />
    </div>
  </form>
</div>
`;
